### PR TITLE
Support target device or group in data triggers

### PR DESCRIPTION
**Pull request by davidebriani**
This PR adds options in the TriggerEditor to be able to specify a target device or target group even for data triggers.
This feature corresponds to setting a specific `device_id` or `group_name` in a Trigger's JSON form when its `type` equals `data_trigger`. For more details, see [the documentation](https://docs.astarte-platform.org/1.0/060-triggers.html#data-triggers).

### CODE REVIEW

File: src/elm/Types/DataTrigger.elm [hunk 1]
```diff
@@ -255,14 +280,29 @@ decoder =
...
 
 
-toDecoder : String -> Int -> DataTriggerEvent -> String -> Operator -> Maybe ( String, JsonType ) -> Decoder DataTrigger
-toDecoder iName iMajor on path operator maybeKnownValue =
-    case maybeKnownValue of
-        Just ( value, jType ) ->
-            Decode.succeed (DataTrigger iName iMajor on path operator value jType)
+toDecoder : String -> String -> String -> Int -> DataTriggerEvent -> String -> Operator -> Maybe ( String, JsonType ) -> Decoder DataTrigger
+toDecoder groupName deviceId iName iMajor on path operator maybeKnownValue =
+    let
```
Comment (Pavinati) [inline src/elm/Types/DataTrigger.elm:285]: didn't test it, but this is the way to go
```
let
    resultValue =
        case maybeKnownValue of
            Just ( value, jType ) ->
                Ok (value, jType)

            Nothing ->
                Err "some error....."

    resultTarget =
        case ( groupName, deviceId ) of
            ( "*", "*" ) ->
                Ok AllDevices

            ( _, "*" ) ->
                Ok (DeviceGroup groupName)

            ( "*", _ ) ->
                Ok (SpecificDevice deviceId)

            ( _, _ ) ->
                Err "Cannot have both device ID and group name"
in
case ( resultValue, resultTarget ) of
    ( Ok (knownValue, knownValueType), Ok target ) ->
        Decode.succeed (DataTrigger iName iMajor on path operator knownValue knownValueType target)

    ( Err msg, _ ) ->
        Decode.fail msg

    ( _, Err msg ) ->
        Decode.fail msg
```

**Review by Pavinati — CHANGES_REQUESTED** [review]

File: src/elm/Types/DataTrigger.elm (hunk 1 above)
Comment (davidebriani) [reply src/elm/Types/DataTrigger.elm:285]: Nice, that's more like it. I admit I struggled a bit with Elm not knowing how to properly exploit conditionals/pattern matching.
I directly copied your snippet and it works fine, thanks!

File: src/elm/Page/TriggerBuilder.elm [hunk 2]
```diff
@@ -1670,6 +1766,106 @@ isValidKnownValue maybeType value =
...
+                            , Input.value deviceId
+                            , Input.onInput UpdateDataTriggerDeviceId
+                            ]
+                        ]
+                    ]
+                ]
+
+
+renderDataTriggerTargetSelect : Bool -> DataTrigger.Target -> Html Msg
+renderDataTriggerTargetSelect editMode target =
```
[the remark this answers is not in the document]
Comment (davidebriani) [reply src/elm/Page/TriggerBuilder.elm:1814]: I tried to extract the methods like you did in your snippet. However the resulting code just seems more convoluted, not to mention I still need to pattern match on `DataTrigger.Target` types with related arguments.
Therefore I would leave the code as it is, if you're ok with it.

[the remark this answers is not in the document]
Comment (Pavinati) [reply src/elm/Page/TriggerBuilder.elm:1814]: We're removing this piece of code soon anyway, so it's ok :+1: 
Plus it's just 3 options. The code is still pretty easy to read

**Review by Pavinati — APPROVED** [review]

**Review by bettio — APPROVED** [review]

**Pull request merged**